geopandas GeoDataFrame support

iboates opened the pull request:
This PR is not expected to be merged but is rather a place to get feedback on implementation of geopandas GeoDataFrames as suggested in https://github.com/networkx/networkx/issues/3932

I have made a humble initial addition which piggybacks off the existing [`from_pandas_edgelist`](https://github.com/networkx/networkx/blob/9b02800087d6a400ce84f9e44a15eee41c27c4dd/networkx/convert_matrix.py#L193) functionality, as well as a few initial tests to verify that it is in fact creating Graph and DiGraph objects

It was not hard to implement this so I am encouraged to continue developing it. In my opinion this could completely usurp the [current shapefile I/O functionality](https://github.com/networkx/networkx/blob/9b02800087d6a400ce84f9e44a15eee41c27c4dd/networkx/readwrite/nx_shp.py). I offer the following advantages to this approach:

1. It allows for NX users to load any geospatial data format (shapefile or one of the many other alternatives) with a well-established, pandas-like library and go directly to a NX graph with little effort.
2. File I/O is handled by geopandas, which leverages other libraries to read and write from/to shapefile and the aforementioned other alternatives, and is therefore no longer NX's concern. As long as you can open it with geopandas, you can make it into a NX graph (if it is a valid network of course)
3. The current implementation is using a very old version of GDAL (1.10.0, two major releases behind the current 3.1.2). I believe it is more prudent to deprecate this functionality and use this nice, well-established and pandas-like library in its place.
4. Geopandas is a fundamentally geospatial library while NX (at least in my humble opinion) is not. GDAL is notoriously difficult to wrangle because it requires system binaries. I think it would be best to focus on graph functionality and leave geospatial implementation details to a library which specializes in them.

I've also been doing some thinking about what exactly needs to be done to make this into a robust functionality. Below are specific functionalities which I think are important to consider:

- [x] Load a GeoDataFrame into a Graph and DiGraph
- [x] Export a Graph or DiGraph to a GeoDataFrame
- [ ] Enable handling of geometric attributes during edge manipulation - This one may be tricky because the physical geometry of the edge is stored as an attribute. So if you flip the orientation of the edge, I would argue that the [LineString](https://shapely.readthedocs.io/en/latest/manual.html#linestrings) object representing the geometry should be flipped correspondingly, so that future export to GeoDataFrame will reflect the changes
- [ ] Handle nodes - This one is also tricky, if you want to have attributes on nodes then you will need to have another GeoDataFrame containing [Point](https://shapely.readthedocs.io/en/latest/manual.html#Point) geometries. The same goes for exporting
- [x] Add a `DeprecationWarning` indicating that support for reading directly from shapefiles will end soon
- [ ] Swap out the `read_shp` function for a legacy-support wrapper which opens the shapefile as a GeoDataFrame and processes it automatically in the meantime.
- [x] Support 3D coordinates
- [x] Handle empty GeoDataFrame and Graph/DiGraph objects
- [ ] Handle mixed 2D & 3D geometries (maybe a `flatten` parameter (default False) which will flatten 3D geometries to 2D in case the user doesn't care about 3D but it's causing problems)
- [ ] Add a few gallery examples

This list is not meant to be exhaustive and I welcome and encourage feedback.

pep8speaks commented:
Hello @iboates! Thanks for updating this PR. 

* In the file [`networkx/convert_matrix.py`](https://github.com/networkx/networkx/blob/b11c4d99f05e2ca5db5417b6e080ea95ed722a0d/networkx/convert_matrix.py):

> [Line 437:89](https://github.com/networkx/networkx/blob/b11c4d99f05e2ca5db5417b6e080ea95ed722a0d/networkx/convert_matrix.py#L437): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 88 characters)
> [Line 438:89](https://github.com/networkx/networkx/blob/b11c4d99f05e2ca5db5417b6e080ea95ed722a0d/networkx/convert_matrix.py#L438): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 88 characters)
> [Line 439:89](https://github.com/networkx/networkx/blob/b11c4d99f05e2ca5db5417b6e080ea95ed722a0d/networkx/convert_matrix.py#L439): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (118 > 88 characters)
> [Line 441:89](https://github.com/networkx/networkx/blob/b11c4d99f05e2ca5db5417b6e080ea95ed722a0d/networkx/convert_matrix.py#L441): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (135 > 88 characters)
> [Line 448:89](https://github.com/networkx/networkx/blob/b11c4d99f05e2ca5db5417b6e080ea95ed722a0d/networkx/convert_matrix.py#L448): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (142 > 88 characters)
> [Line 449:89](https://github.com/networkx/networkx/blob/b11c4d99f05e2ca5db5417b6e080ea95ed722a0d/networkx/convert_matrix.py#L449): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 88 characters)
> [Line 452:89](https://github.com/networkx/networkx/blob/b11c4d99f05e2ca5db5417b6e080ea95ed722a0d/networkx/convert_matrix.py#L452): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (130 > 88 characters)
> [Line 453:89](https://github.com/networkx/networkx/blob/b11c4d99f05e2ca5db5417b6e080ea95ed722a0d/networkx/convert_matrix.py#L453): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (132 > 88 characters)
> [Line 473:89](https://github.com/networkx/networkx/blob/b11c4d99f05e2ca5db5417b6e080ea95ed722a0d/networkx/convert_matrix.py#L473): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (146 > 88 characters)
> [Line 479:89](https://github.com/networkx/networkx/blob/b11c4d99f05e2ca5db5417b6e080ea95ed722a0d/networkx/convert_matrix.py#L479): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (114 > 88 characters)
> [Line 496:89](https://github.com/networkx/networkx/blob/b11c4d99f05e2ca5db5417b6e080ea95ed722a0d/networkx/convert_matrix.py#L496): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (94 > 88 characters)

* In the file [`networkx/tests/test_convert_geopandas.py`](https://github.com/networkx/networkx/blob/b11c4d99f05e2ca5db5417b6e080ea95ed722a0d/networkx/tests/test_convert_geopandas.py):

> [Line 22:89](https://github.com/networkx/networkx/blob/b11c4d99f05e2ca5db5417b6e080ea95ed722a0d/networkx/tests/test_convert_geopandas.py#L22): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (108 > 88 characters)
> [Line 34:89](https://github.com/networkx/networkx/blob/b11c4d99f05e2ca5db5417b6e080ea95ed722a0d/networkx/tests/test_convert_geopandas.py#L34): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (110 > 88 characters)
> [Line 46:89](https://github.com/networkx/networkx/blob/b11c4d99f05e2ca5db5417b6e080ea95ed722a0d/networkx/tests/test_convert_geopandas.py#L46): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (110 > 88 characters)
> [Line 52:89](https://github.com/networkx/networkx/blob/b11c4d99f05e2ca5db5417b6e080ea95ed722a0d/networkx/tests/test_convert_geopandas.py#L52): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (96 > 88 characters)
> [Line 54:89](https://github.com/networkx/networkx/blob/b11c4d99f05e2ca5db5417b6e080ea95ed722a0d/networkx/tests/test_convert_geopandas.py#L54): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (99 > 88 characters)
> [Line 60:12](https://github.com/networkx/networkx/blob/b11c4d99f05e2ca5db5417b6e080ea95ed722a0d/networkx/tests/test_convert_geopandas.py#L60): [W292](https://duckduckgo.com/?q=pep8%20W292) no newline at end of file

Please install and run [psf/black](https://github.com/psf/black).

iboates commented:
You can now export a graph to a geopandas GeoDataFrame, but I am realizing that it is re-using a large amount of the code from from_pandas_edgelist() and to_pandas_edgelist(). What would you think about simply giving these two already-established functions a `geoemtry` parameter, that, when not set to a default `None` value, would trigger the export of a GeoDataFrame with an attribute named after the edge's geometry attribute?

jarrodmillman commented:
Thanks for working on this!  If I am correctly understanding you, this would provide a nice path forward for us and allow us to drop the dependency on the GDAL/OGR libraries.  I am going to summarize what I think you are saying to make sure I am not confused.

You are proposing that if NX users want to use geospatial data with NX, they would use Fiona via geopandas to handle file I/O.  Then the only thing about geospatial data that NX needs to be aware of is how to convert geopandas GeoDataFrames to NX graphs and vice-versa, which is what ``from_geopandas_edgelist`` and ``to_geopandas_edgelist`` provide.

If so, this could be how we roll out the change:

1. NX 2.5 (Sept 2020) would add ``from_geopandas_edgelist`` and ``to_geopandas_edgelist`` as well as deprecate everything in ``readwrite/nx_shp.py``.

2. NX 3.0 (Jan 2021) would remove ``readwrite/nx_shp.py`` and at that point we could remove gdal from our dependencies.

I removed ``gdal-libs`` and ``gdal-devel`` from my system and was able to pip install ``geopandas``, ``Fiona``, and ``Shapely``.  So it seems that they don't need you to install the GDAL/GR libraries.  Is that correct?  Does the wheel bundle the GDAL/OGR libraries?

However, I am a bit confused b/c according to https://fiona.readthedocs.io/en/latest/README.html#installation
```
Fiona requires Python versions 3.6+ and GDAL version 1.11-3.0.
```
And later it says
```
Assuming you’re using a virtualenv (if not, skip to the 4th command) and GDAL/OGR libraries, headers, and gdal-config program are installed to well known locations on your system via your system’s package manager (brew install gdal using Homebrew on OS X), installation is this simple.

$ mkdir fiona_env
$ virtualenv fiona_env
$ source fiona_env/bin/activate
(fiona_env)$ pip install fiona
```
That makes me think you do need the GDAL/OGR libraries installed.  But I explicitly removed the ``GDAL/OGR libraries, headers, and gdal-config program``, created a virtual environment, and pip installed ``geopandas``:
```
$ sudo dnf remove gdal-devel gdal-libs
$ python3 -m venv $VENV_HOME/test
$ source "$VENV_HOME/test/bin/activate"
(test) $ pip install geopandas
(test) $ pip list
Package         Version     
--------------- ------------
attrs           19.3.0      
click           7.1.2       
click-plugins   1.1.1       
cligj           0.5.0       
Fiona           1.8.13.post1
geopandas       0.8.1       
munch           2.5.0       
numpy           1.19.1      
pandas          1.0.5       
pip             19.3.1      
pyproj          2.6.1.post1 
python-dateutil 2.8.1       
pytz            2020.1      
setuptools      41.6.0      
Shapely         1.7.0       
six             1.15.0      
```
It seems there is no issue installing Fiona, Shapely, and geopandas without having to install the GDAL/OGR libraries.  Am I missing something?

If not, do my steps 1 and 2 above make sense?

We will probably need to create a NXEP (https://networkx.github.io/documentation/latest/developer/nxeps/index.html) before we can decide how to move forward.  That would be a bit down the road after we've had a little more discussion and I am more than happy to help you with that.  It would also be worth creating some gallery or tutorial examples illustrating how to use the new functionality as part of the PR once we've made more progress on moving it foward.

PS.  We don't actually depend on gdal==1.10.0.  That is in the requirements file, but we test on gdal==3.1.2 and it works fine (although with a different version of the GDAL library).  The issue I've had is that every time I use a new OS, I have to use different fixes to install the GDAL library.  I've been hoping we could remove the dependency on gdal and the GDAL/OGR libraries, so I haven't been sure what to do about our gdal requirement.

iboates commented:
@jarrodmillman I think what happens is that fiona gets installed, but it then does not have access to the GDAL/OGR system binaries and therefore is not able to read or write files properly. Maybe you could try making this command in your configuration:

    >>> import fiona
    >>> fiona.supported_drivers

I got this from the documentation for file I/O in geopandas:

https://github.com/geopandas/geopandas/blob/d3228ddb6814515506921ad5c2e6d6f8bd99c9ae/geopandas/io/file.py#L172

You could also try the new tests I created in this PR, which will use geopandas to open a shapefile, create a geodataframe, create a graph from the dataframe, and also export it back to a dataframe. Furthermore you could try to write the geodataframe to a file by calling

    gdf.to_file("path/to/file.shp")

And see if it correctly exports a shapefile (It should actually export many files because shapefiles are an old composite file type consisting of several different files).

jarrodmillman commented:
That is unfortunate.  My main concern is getting rid of the dependency on the GDAL/OGR libraries for networkx.

Since networkx is a pure Python package it is a very minimal dependency (especially if we get rid of the ``gdal`` requirement).  It would be much easier for geopandas (or Fiona) to require networkx than for networkx to require geopandas, which requires fiona (which requires the GDAL/OGR libraries).  Is there any reason that the functionality of ``from_geopandas_edgelist`` and ``to_geopandas_edgelist`` shouldn't be added to geopandas (or Fiona instead of networkx)?

jarrodmillman commented:
Have you looked at https://github.com/caesar0301/s2g?

Another idea, would be to create a new project
- ``https://github.com/networkx/gpd2nx``

If we decide to do that I would setup the packages and make releases.

jarrodmillman commented:
@iboates The more I think about this, the more I like it.  I fixed the test failures
and added the functions to our documentation.  You will need to use the numpy
docstring standard for this to render properly, though.

We recently started autobuilding the documentation for PRs, so I rebased to master
and forced pushed to your branch.

You can see the documentation here:
https://15-890377-gh.circle-artifacts.com/0/doc/build/html/reference/convert.html?highlight=geopandas

Git will be confused on your local machine because I changed history
when I force pushed to your remote branch.  So you will have to run
a few commands.  If you aren't familiar with what to do, please take
a look at this:
https://stackoverflow.com/questions/9813816/git-pull-after-forced-update

If you aren't sure after looking at that, please don't hesitate to ask.

I also formatted the code using https://github.com/psf/black

You may want to take a look at our contributor guide:
https://networkx.github.io/documentation/latest/developer/contribute.html
In particular, near the top it shows you how to install a pre-commit hook that will format the code for you.

There is still one error that needs to be fixed.  Do you know how to fix it?

iboates commented:
@jarrodmillman I wanted to come and reply to your post before you edited it, but apparently silence has proven to be the most compelling argument!

But seriously, what you said made me reflect on and re-consider my own opinion as well, and for the good of the project I think it's still worth saying, so here it goes:

I don't think it will be possible to remove GDAL and still retain the ability to manipulate geometric/geographic features. At least, not without writing your own geometric manipulation libs. It is basically bedrock for all GIS applications that I know of, except maybe in some Java ecosystems.

If `nx.from_pandas_edgelist()` is able to accept `pd.DataFrame` objects, then it should technically also be able to accept `gpd.GeoDataFrame` objects, as long as they have source and target columns. This is because they inherit directly from `pd.DataFrame`, and only have extra semantic, modifications around them for strictly geospatial properties and methods. So the argument could be made that generating an network-compatible edgelist (i.e. having a source & target column could actually be a responsibility of geopandas rather than networkx.

That argument, however, leaves networkx in a bit of an awkward position, because `nx.read_shp()` still exists, effectively being grandfathered due to the fact that it has simply existing longer than geopandas rise to notoriety (I think). So then it begs the question - if turning the `gpd.GeoDataFrame` into a networkx-compatible edgelist now lies with geopandas, should `nx.read_shp()` even exist anymore? And if the answer is "no" then you are removing a very useful functionality and telling users that they have to now go install a completely different set of libs just to get the same functionality back.

So, having expressed that, if you still like the addition then I will continue working on it. Currently I'm not totally satisfied with how the nodes are parsed out of the `LineString` objects because it uses `.iterrows()`, which is generally a faux-pas in the pandas world due to its inefficiency. But thus far I have not been able to find a preferred `.apply()`-based alternative.

dschult commented:
I think ```.itertuples``` is supposed to be an improved ```.iterows``` but I haven’t looked closely at this case.

iboates commented:
@dschult Indeed. I just did some testing  between `.itertuples()`, `.iterrows()` and a custom solution where I used a static class to manage node creation & id assignment and `.itertuples()` is the clear winner. In smaller GeoDataFrames the gains are negligible (as expected) but when I tried with a GeoDataFrame with ~40k rows, the processing time was nearly halved.

iboates commented:
@jarrodmillman I have installed the pre-commit hooks.

What was the error that you said needs to still be fixed? When I run pytest in the tests dir I only get is from `test_from_numpy_matrix_type`, and that seems to just be due to an `ImportError` caused by scipy not being installed.

pep8speaks commented:
Hello @iboates! Thanks for updating this PR. 

* In the file [`networkx/convert_matrix.py`](https://github.com/networkx/networkx/blob/188117b7afc7a73dc1e6a4f37d73dbb70c6d36fa/networkx/convert_matrix.py):

> [Line 441:89](https://github.com/networkx/networkx/blob/188117b7afc7a73dc1e6a4f37d73dbb70c6d36fa/networkx/convert_matrix.py#L441): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (98 > 88 characters)
> [Line 478:89](https://github.com/networkx/networkx/blob/188117b7afc7a73dc1e6a4f37d73dbb70c6d36fa/networkx/convert_matrix.py#L478): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (110 > 88 characters)
> [Line 479:89](https://github.com/networkx/networkx/blob/188117b7afc7a73dc1e6a4f37d73dbb70c6d36fa/networkx/convert_matrix.py#L479): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (104 > 88 characters)
> [Line 491:89](https://github.com/networkx/networkx/blob/188117b7afc7a73dc1e6a4f37d73dbb70c6d36fa/networkx/convert_matrix.py#L491): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (142 > 88 characters)
> [Line 500:89](https://github.com/networkx/networkx/blob/188117b7afc7a73dc1e6a4f37d73dbb70c6d36fa/networkx/convert_matrix.py#L500): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (97 > 88 characters)
> [Line 505:89](https://github.com/networkx/networkx/blob/188117b7afc7a73dc1e6a4f37d73dbb70c6d36fa/networkx/convert_matrix.py#L505): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (122 > 88 characters)
> [Line 509:89](https://github.com/networkx/networkx/blob/188117b7afc7a73dc1e6a4f37d73dbb70c6d36fa/networkx/convert_matrix.py#L509): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (125 > 88 characters)

* In the file [`networkx/readwrite/nx_shp.py`](https://github.com/networkx/networkx/blob/188117b7afc7a73dc1e6a4f37d73dbb70c6d36fa/networkx/readwrite/nx_shp.py):

> [Line 85:89](https://github.com/networkx/networkx/blob/188117b7afc7a73dc1e6a4f37d73dbb70c6d36fa/networkx/readwrite/nx_shp.py#L85): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (242 > 88 characters)

* In the file [`networkx/readwrite/tests/test_shp.py`](https://github.com/networkx/networkx/blob/188117b7afc7a73dc1e6a4f37d73dbb70c6d36fa/networkx/readwrite/tests/test_shp.py):

> [Line 222:13](https://github.com/networkx/networkx/blob/188117b7afc7a73dc1e6a4f37d73dbb70c6d36fa/networkx/readwrite/tests/test_shp.py#L222): [F841](https://duckduckgo.com/?q=pep8%20F841) local variable 'G' is assigned to but never used

* In the file [`networkx/tests/test_convert_geopandas.py`](https://github.com/networkx/networkx/blob/188117b7afc7a73dc1e6a4f37d73dbb70c6d36fa/networkx/tests/test_convert_geopandas.py):

> [Line 67:89](https://github.com/networkx/networkx/blob/188117b7afc7a73dc1e6a4f37d73dbb70c6d36fa/networkx/tests/test_convert_geopandas.py#L67): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (99 > 88 characters)



Please install and run [psf/black](https://github.com/psf/black).

iboates commented:
I've made a few additions:

- Added support for 2D and 3D geometries in GeoDataFrames
- Added `DeprecationWarning` to `nx.read_shp`
- Switched from `.iterrows` to `.itertuples`
- Updated docstrings to numpy standard
- Added tests for empty GeoDataFrames & Graphs/DiGraphs

We could technically get rid of the shapefiles that I have added. The test GeoDataFrames could be built by manually constructing their constituent LineString geometries in the test code itself. This might be a good idea because one of the goals of this is to make geographic file I/O not anymore NX's responsibility. But maybe you would like to keep them just to be sure that the end-to-end user story (so to speak) works?

jarrodmillman commented:
For the errors, check the CI builds.  At the bottom of the page, you will see red ``x``s by failed tests.  Click on them and investigate.  For some reason, Travis is no longer being run.  But, you will probably still need to fix those errors.  Hopefully, it will come back when you make more commits (so you can see the errors).  Your recent changes broke the tests on Appveyor, so I would start there.  I took a quick look and you probably just need to make an entry for ``geopandas`` similar to the one for ``pandas`` in ``networkx/conftest.py`` so that ``pytest`` doesn't try to run tests that depend on ``geopandas`` when it isn't available.

I am not sure why black isn't being run on your computer.  To confirm that you installed it, try looking at the output of ``pip list``.  If you are having trouble using the pre-commit hook, you can run black manually.  For instance, to run it on ``networkx/convert_matrix.py``:
```
black networkx/convert_matrix.py
```
You will need to run it on all the Python files you edited.

Could you also add a note to the bottom of ``doc/developer/deprecations.rst`` about the deprecations.

jarrodmillman commented:
This is a fairly big change, since we are dropping ``nx.read_shp()`` and ``nx.write_shp()``.  So it would be good if @dschult and @rossbar could comment on this more.  In particular, do you agree this is the correct way forward?  It would also be great if you could review the code at some point.

jarrodmillman commented:
@iboates I added ``Add a few gallery examples`` to your checklist.  It may make sense to add a new directory (maybe ``geospatial``) to ``networkx/examples``.  You should be able to just follow the other examples.  We use
- https://sphinx-gallery.github.io/stable/index.html
to build the gallery.  So you may find the ``sphinx-gallery`` documentation helpful.  Also feel free to ask if you have any questions.

dschult commented on networkx/convert_matrix.py:
```diff
@@ -425,6 +427,203 @@ def from_pandas_edgelist(
+
+
+def to_geopandas_edgelist(
+    G, source="source", target="target", geometry="geometry", crs=None, nodelist=None,
```
```geometry``` and ```crs``` should be added to the doc_string of ```to_geopandas_edgelist```.
Perhaps you should change the order of the parameters to list ```nodelist``` before ```geometry``` or ```crs```?

dschult replied:
Also, the ```nodelist``` description should be tweaked a little. The edges that are added have starting node in ```nodelist```. It is not true that edges must include both incident nodes in ```nodelist```. 

Perhaps "Use only nodes specified in nodelist" -> "Include only edges that start at one of the nodes in nodelist"

dschult commented on networkx/tests/test_convert_geopandas.py:
```diff
@@ -0,0 +1,85 @@
+    failed = False
+    for u, v, d in graph_from_shp.edges(data=True):
+        if set(d.keys()) != {"id", "attr_text", "attr_float", "geometry", "attr_int"}:
+            failed = True
```
You can put the assert statement in the loop. It will fail upon the first non-equal case just as this code does.
Similarly for the directed case below.

dschult commented on networkx/tests/test_convert_geopandas.py:
```diff
@@ -0,0 +1,85 @@
+        )
+        # If length of set is not 1 then there is True and False
+        # If The first length of set is 1 but the next iterated item is False then there is no True
+        if len(set(joined["test"])) > 1 or not next(iter(set(joined["test"]))):
```
I think this if statement could be: ```if True in joined["test"]```
But it might be more compact to make the loop be:
```python
for col in [c for c in joined.columns if c[-4:] != "_new"]:
    assert not any(joined.apply(
            lambda row: row[col] == row[f"{col}_new"], axis="columns"
        ))
```

dschult commented on networkx/tests/test_convert_geopandas.py:
```diff
@@ -0,0 +1,85 @@
+    assert not failed
+
+
+def test_to_geopandas_edgelist():
```
This test has the potential to check that ```to_geopandas_edgelist``` and ```from_geopandas_edgelist``` are inverses -- that is, the ```gdf``` we start with equals the ```new_gdf``` we end with after a round-trip into nx.
You check that the data is the same, but not that the edges are the same. 
Is there an easy: ```gdf == new_gdf``` capability in geopandas?  Something similar?

dschult commented:
You should add the deprecation warning to ```networkx/conftest.py``` so it gets ignored. Also in that file, you'll need to add code so Appveyor can test on Windows when geopandas is not present. I would suggest using the pandas dependency as a template for your geopandas dependency in that file.

iboates commented:
@dschult 

> Finally, the "big picture" question: Can this "just" be done with to/from_pandas_edgelist by including a column named "geometry" in the pandas DataFrame (or in the edge attributes)? Please justify why this should be an extra set of functions rather than an example showing people how to use a GeoDataFrame with the previously available functions. [And Thanks for All This!]

This is something I have been internally wrestling with. Theoretically, a GeoDataFrame should be able to be handled just fine by `from_pandas_edgelist` because it inherits from `pd.DataFrame`. Therefore, the argument could be made that the `geopandas` library could be modified to have some kind of `make_network` function which would generate source & target columns which would make it compatible with `nx.from_pandas_edgelist`.

The advantage to this approach is that the detection and marking of geometric line endpoint overlaps would then fall completely within the jurisdiction of `geopandas`, which already carries the burden of managing its own dependencies. The disadvantage is that it doesn't change anything about `nx.read_shp`. It could be phased out in favour of using `geopandas`, but then it is a lost functionality rather than an upgraded one.

I'm sorry to second-guess my own proposal but the way forward doesn't seem all that clear to me. I think there should be a consensus among the core maintainers about this before I work any more on the enhancement.

But I do agree with all the other specific comments you made.

jarrodmillman commented:
I am not certain what the way forward is, so I am just thinking out loud.

1. We decided to just release 2.5 now and release a 2.6 in December before 3.0 in early 2021.  That gives us some time to figure out the best way forward.

2. The core developers don't use or know much about geospatial data.  From our perspective, ``nx.read_shp`` is one of the most difficult bits of code to maintain.  I spend more time dealing with ``gdal`` than any other dependency.  So I am motivated to remove our dependency on it somehow.

3. We have no documentation or gallery examples using geospatial data.

Here is a possible way forward that would deprecate ``read_/write_shp`` in 2.6 and remove it in 3.0.

If geopandas gets a ``make_network`` function, we could work with you to develop several geospatial examples in our gallery:

  https://networkx.github.io/documentation/latest/auto_examples/index.html

I am imaging a section called ``geospatial`` containing several small examples with data, text, code, and figures.  All the examples would use geopandas and ``from_pandas_edgelist``.  We use sphinx-gallery

https://github.com/sphinx-gallery/sphinx-gallery/

to make the gallery from python scripts

https://github.com/networkx/networkx/tree/master/examples

We are planning to improve our gallery and this could be a first example.  We would also provide you with help developing the examples, but we don't know anything about geospatial data.  So you will need to supply the ideas.

We would also be happy to help you and the geopandas developers with the ``make_network`` function.  We would set up our CI system to install new releases of geopandas so that we would be able to test that nothing breaks.  geopandas could also add an example of using networkx to test things on their end.  We already do this with scikit-image, since they depend on networkx.  Whenever I make a networkx prerelease I watch the scikit-image CI system to make sure we didn't break anything they depend on.

rossbar commented:
Having read through the discussion here and in #3932 I like the general approach laid out by @jarrodmillman . Having never worked with geospatial data, it seems it would make sense for I/O with shp files (and other domain-specific file formats) to live under a different umbrella. @iboates I agree with your concern about losing functionality, but having a viable alternative that is a) well-documented (e.g. with gallery examples) and b) depends only on well-established tools in the geospatial ecosystem (I assume geopandas fits this bill) should mitigate problems in the short term and be much better for both libraries in the long run. Additionally, a noisy deprecation warning for `nx._shp` gives users plenty of opportunity to push back - if it turns out that removing the shp file I/O from nx is too disruptive then we can change tacks. Also thanks for the PR @iboates , the provided code really helps frame the discussion!

iboates commented:
I just returned from holiday and have read through the last few comments. Thanks for the feedback everyone. I will propose adding this functionality to the geopandas library in the form of a function which will build a line topology using parts of the code in this PR, thereby making it compatible with `nx.from_pandas_edgelist`.

jarrodmillman commented:
@iboates Please reference this ticket when you propose adding this functionality to geopandas, so we can comment and offer help and answer questions about working with nx.

jarrodmillman commented:
@iboates NetworkX is deprecating read/write functionality for shapefiles (#4298) and providing gallery examples for working with geospatial data using the main packages for geospatial data analysis for Python (#4366). We are interested in making sure our gallery examples provide the recommended way to read/write geospatial data and converting them back and forth to NetworkX. We would love feedback and additional gallery examples! One of our main goals is to make it clear to our users how to get started and where to go for questions regarding geospatial data formats since we (the NetworkX developers) don't work with geospatial data and don't know how to provide guidance for users.

jarrodmillman commented:
See #4298 and #4366.
